app/jobs/send_gpo_code_expiration_notices_job.rb

```diff
@@ -0,0 +1,77 @@
+class SendGpoCodeExpirationNoticesJob < ApplicationJob
+  queue_as :low
+
+  def initialize(analytics: nil)
+    @analytics = analytics
+  end
+
+  def perform
+    codes_to_send_notifications_for.find_each do |code|
+      user = code.profile.user
+
+      user.send_email_to_all_addresses(:gpo_code_expired, code_sent_at: code.code_sent_at)
+      code.update(expiration_notice_sent_at: Time.zone.now)
+
+      analytics.idv_gpo_expiration_email_sent(user_id: user.uuid)
+    end
+  end
+
+  def codes_to_send_notifications_for
+    # We are looking at a 48 hr window in which all codes will _definitely_ be expired.
+    from, to = calculate_notification_window_bounds
+    expired_codes_needing_notification_sent_between(
+      from: from,
+      to: to,
+    )
+  end
+
+  def calculate_notification_window_bounds(as_of: Time.zone.now)
+    to = as_of.beginning_of_day - IdentityConfig.store.usps_confirmation_max_days.days
+    from = to - 2.days
+    [from, to]
+  end
+
+  private
+
+  def analytics
+    @analytics ||= Analytics.new(user: AnonymousUser.new, request: nil, session: {}, sp: nil)
+  end
+
+  def expired_codes_needing_notification_sent_between(
+    from:,
+    to:
+  )
+    GpoConfirmationCode.joins(:profile).
+      # 1. Exclude codes that we've already sent an expiration notice for
+      where(expiration_notice_sent_at: nil).
+
+      # 2. Exclude codes not sent in the window we're looking at
+      where(code_sent_at: from...to).
```

Comment:
+1 to just passing around the `time_range` as a single range
```suggestion
      where(code_sent_at: time_range).
```